tc_build/kernel.py

```diff
@@ -0,0 +1,401 @@
+
+        super().build()
+
+    # https://github.com/llvm/llvm-project/commit/33504b3bbe10d5d4caae13efcb99bd159c126070
+    def can_use_ias(self):
+        return self.toolchain_version >= (14, 0, 2)
+
+    # https://github.com/ClangBuiltLinux/linux/issues/1601
+    def needs_binutils(self):
+        return True
+
+
+class RISCVKernelBuilder(KernelBuilder):
+
+    def __init__(self):
+        super().__init__('riscv')
+
+        self.cross_compile = 'riscv64-linux-gnu-'
+
+    # https://github.com/llvm/llvm-project/commit/bbea64250f65480d787e1c5ff45c4de3ec2dcda8
+    def can_use_ias(self):
+        return self.toolchain_version >= (13, 0, 0)
+
+
+class S390KernelBuilder(KernelBuilder):
+
+    def __init__(self):
+        super().__init__('s390')
+
+        self.cross_compile = 's390x-linux-gnu-'
+
+        # LD: https://github.com/ClangBuiltLinux/linux/issues/1524
+        # OBJCOPY: https://github.com/ClangBuiltLinux/linux/issues/1530
+        # OBJDUMP: https://github.com/ClangBuiltLinux/linux/issues/859
+        for key in ['LD', 'OBJCOPY', 'OBJDUMP']:
+            self.make_variables[key] = self.cross_compile + key.lower()
+
+    def build(self):
+        self.toolchain_version = self.get_toolchain_version()
+        if self.toolchain_version <= (15, 0, 0):
+            # https://git.kernel.org/linus/30d17fac6aaedb40d111bb159f4b35525637ea78
+            utils.print_warning('s390 does not build with LLVM < 15.0.0, skipping build...')
+            return
+
+        super().build()
+
+    def can_use_ias(self):
+        return True
+
+    def needs_binutils(self):
+        return True
+
+
+class X8664KernelBuilder(KernelBuilder):
+
+    def __init__(self):
+        super().__init__('x86_64')
+
+
+class LLVMKernelBuilder(Builder):
+
+    def __init__(self):
+        super().__init__()
+
+        self.bolt_instrumentation = False
+        self.bolt_sampling_output = None
+        self.matrix = {}
+        self.toolchain_prefix = None
+
+    def build(self):
+        builders = []
+
+        allconfig_capable_builders = {
+            'AArch64': Arm64KernelBuilder,
+            'ARM': ArmKernelBuilder,
+            'Hexagon': HexagonKernelBuilder,
+            'RISCV': RISCVKernelBuilder,
+            'SystemZ': S390KernelBuilder,
+            'X86': X8664KernelBuilder,
+        }
+
+        # This is a little convoluted :/
+        # The overall idea here is to avoid duplicating builds, so the
+        # matrix consists of a series of configuration targets ("defconfig",
+        # "allmodconfig", etc) and a list of LLVM targets to build for each
+        # configuration target. From there, this block filters out the
+        # architectures that cannot build their "all*configs" with clang, so
+        # they are duplicated if both "defconfig" and "allmodconfig" are
+        # requested.
+        for config_target, llvm_targets in self.matrix.items():
+            if config_target == 'defconfig':
```

Comment:
To avoid more and more indentation, when there's no else clause you could always invert the conditional and continue early.

```python
if config_target != 'defconfig':
    continue
for llvm_target in llvm_targets:
```
that way you code doesn't keep creeping to the right.

Reply:
I do not think that suggestion would work for what this block is trying to do but I tried to simplify it in 0c1452ee9c5ea3d21ecfb5c08b682b75a9610f9f to make it clearer, which reduces the indentation.